hal/tests/TESTS/pin_names/pinvalidate.py

```diff
@@ -46,286 +79,392 @@ def find_target(target_name = ''):
+                if case["suite_name"] != suite:
                     continue
-                if case['result'] == 'FAILED':
-                    result = 'FAILED'
+                if case["result"] == "FAILED":
+                    result = "FAILED"
                     error_count += 1
             table.append([target, suite, result, error_count])
-    
-    return tabulate(table, headers=['Platform name', 'Test suite', 'Result', 'Error count'], tablefmt='grid')
 
-def print_report(report, print_error_detail, tablefmt='grid'):
+    return tabulate(
+        table,
+        headers=["Platform name", "Test suite", "Result", "Error count"],
+        tablefmt="grid",
+    )
+
+
+def print_report(report, print_error_detail, tablefmt="grid"):
     table = []
     for case in report:
         errors_str = ""
-        for error in case['errors']:
-            errors_str += '\n'
-            if error['key']:
-                errors_str += error['key']
-            if error['val']:
-                errors_str += " = " + error['val']
-            if error['message']:
-                errors_str += " <-- " + error['message']
-        
-        if errors_str == '':
-            errors_str = 'None'
+        for error in case["errors"]:
+            errors_str += "\n"
+            if error["key"]:
+                errors_str += error["key"]
+            if error["val"]:
+                errors_str += " = " + error["val"]
+            if error["message"]:
+                errors_str += " <-- " + error["message"]
+
+        if errors_str == "":
+            errors_str = "None"
 
         if print_error_detail:
-            table.append([case['platform_name'], case['suite_name'], case['case_name'], case['result'], len(case['errors']), errors_str.lstrip()])
+            table.append(
+                [
+                    case["platform_name"],
+                    case["suite_name"],
+                    case["case_name"],
+                    case["result"],
+                    len(case["errors"]),
+                    errors_str.lstrip(),
+                ]
+            )
         else:
-            table.append([case['platform_name'], case['suite_name'], case['case_name'], case['result'], len(case['errors'])])
+            table.append(
+                [
+                    case["platform_name"],
+                    case["suite_name"],
+                    case["case_name"],
+                    case["result"],
+                    len(case["errors"]),
+                ]
+            )
+
+    return tabulate(
+        table,
+        headers=[
+            "Platform name",
+            "Test suite",
+            "Test case",
+            "Result",
+            "Error count",
+            "Errors",
+        ],
+        tablefmt=tablefmt,
+    )
 
-    return tabulate(table, headers=['Platform name', 'Test suite', 'Test case', 'Result', 'Error count', 'Errors'], tablefmt=tablefmt)
 
 def print_pretty_html_report(report):
```

Comment:
If you are going to be concatenating a lot of strings together, this is potentially really slow. Each time you concatenate a string to another string, you are allocating memory for that new string at that moment and not necessarily deallocating the old right at that point in time. 

A better way would be to make output a list of strings you want like so:
```python
output.append("<p>My first string.</p>\n")
output.append(f"<p>Error count: {err_count}</p>\n")
```
After that, at the end, you can either return the list and iterate over the output (i.e. `file.writelines(iterator)`) or, if you really want that string instead, you can do this:
```python
return "".join(output)
```
which will be faster.

Reply:
Cool, will consider this!

Reply:
Done